### ISSUE TITLE
Why the very first design time build after switching from console to VS IDE changes the *.csproj.AssemblyReference.cache files?

### ISSUE THREAD

**Issue by MarkKharitonov** [dotnet/msbuild#7800]
## Visual Studio Version
17.1.3

## Summary
I check the *.csproj.AssemblyReference.cache files after console msbuild and after the first design time build in VS IDE. They are different even though nothing has changed. Note that consecutive msbuild runs or regular builds in VS IDE do not change them. Only after switching from the console build to a new VS IDE instance that runs the first design time build do we see the issue.

## Steps to Reproduce

1. git clean -qdfx
2. build with msbuild
3. Save the *.csproj.AssemblyReference.cache files
4. build with msbuild
5. Save the *.csproj.AssemblyReference.cache files
6. open devenv
7. wait for the design time build to finish
8. Check the *.csproj.AssemblyReference.cache files 

I have the binary logs and the selected files saved aside for each build including the design time build, but I do not want to upload them to a public location. Is there a private upload link?

## Expected Behavior
All the files are identical.

## Actual Behavior
The design time build produces a different file

## User Impact
I do not know. It may be a symptom of a bigger problem.

**Comment by danmoseley** [comment]
Are the cache files byte for byte identical with the ones they over wrote?

**Comment by MarkKharitonov** [comment]
No, they are different. The code does not recompile, so it does not affect the `Csc` task, but I am afraid it affects the `ResolveAssemblyReferences` task, which is not less expensive.
I have captured the timestamps before and after plus the binary logs themselves. I can rerun it and capture some cache files before and after. If there is a private upload link, I can upload all of it there.

**Comment by danmoseley** [comment]
It might be interesting to binary diff them to get an idea what changed, e.g. something like a version number. (I'm not on the team and no doubt the code is totally changed from when I was)

**Comment by MarkKharitonov** [comment]
Here you go - https://developercommunity.visualstudio.com/t/Why-msbuild-overwrites-some-csprojAss/10090995?space=61&entry=problem

How do I get an upload link?

**Comment by MarkKharitonov** [comment]
![image](https://user-images.githubusercontent.com/489792/178032932-d91045df-2a79-4811-aabc-97f6f931701f.png)

Please, advise.

**Comment by Forgind** [comment]
> Theory: the change in the cache may be a result of running on a different node that has built different projects and thus has different things in its in-memory cache.

If the in-memory cache has anything the file state doesn't, it should write it into the file state, so the file state should be complete either way. Guess pre-looking at it: danmoseley's version change sounds plausible, or it could just be that it found the same version at a different path.

**Comment by Forgind** [comment]
Small update:
I arbitrarily looked at the second assembly that was "different." Apparently the version was the same, but the lastModified timestamp changed—from 3/26/2018 to 3/27/2018. This was Microsoft.CSharp.dll. The paths were the same, as was everything else about them. My current hypothesis is that there's something different about how timestamps are calculated between when we calculate them from a command line build and when VS calculates them. I can probably find how we calculate them, but I will need to solicit help to figure out how VS calculates them.

**Comment by Forgind** [comment]
The two timestamps are four hours apart. @MarkKharitonov, are you four hours away from UTC, by chance?

**Comment by Forgind** [comment]
Ok, so we seem to use UTC. I would argue that's better behavior, so if I can find code suggesting VS uses local time for timestamps, I'd suggest we recommend they change that. (I'm not sure if they'll be able to; it might be a breaking change.)

**Comment by Forgind** [comment]
The other interesting part of this is that it only seemed to affect some packages. If I'm right, that suggests that some packages' timestamps were recorded in (or converted to) local time, whereas others were UTC.

**Comment by MarkKharitonov** [comment]
@Forgind - I am in EST, which is indeed 4 hours off the UTC.

**Comment by danmoseley** [comment]
I see no reason for time stamps to be stored in local time. When I travel, it shouldn't trigger a clean build.

One scenario VS might care about though (assuming it is a VS issue) is operating on the same build tree with both an older and newer VS. Presumably switching to the other one spoils not trigger a rebuild either.
